### PR TITLE
build: clean build script, restrict hoist of pnpm

### PR DESCRIPTION
**Pull request by fi3ework**
## Summary

- restrict pnpm hoisted dependencies to avoid phantom deps for deps.
   - add missing dependencies after hoisting has been restricted.
   - set `cwd` for certain test cases to correctly resolve dependencies after hoisting has been restricted. (By the way, I believe all the current working directories should be set to the case directory instead of the current working directory, which is the workspace root now.)
   - remove the non-pure shims for ESM. Since the phantom dependencies are no longer available, the test case `e2e/cases/mjs-artifact/test.mjs` would fail because `global.require` will be set concurrently and fail due to the non-availability of the phantom dependency.
- fix some steps in the contribution doc.

## Checklist

- [x] Tests updated.
- [x] Documentation updated.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -147,17 +147,17 @@ pnpm run --filter @rsbuild/some-package test
```
Comment (chenjiahan) [inline CONTRIBUTING.md:147]: This PR contains several changes, can you split them to into separate PRs? 😄

Small PR is better for reviewing and generating changelog.

Comment (fi3ework) [reply CONTRIBUTING.md:147]: Sure. I'm closing this PR for now.
